cranelift/interpreter/src/interpreter.rs

```diff
@@ -257,38 +275,93 @@ impl<'a> State<'a, DataValue> for InterpreterState<'a> {
         self.fflags.clear()
     }
 
-    fn load_heap(&self, offset: usize, ty: Type) -> Result<DataValue, MemoryError> {
-        if offset + 16 < self.heap.len() {
-            let pointer = self.heap[offset..offset + 16].as_ptr() as *const _ as *const u128;
-            Ok(unsafe { DataValue::read_value_from(pointer, ty) })
-        } else {
-            Err(MemoryError::InsufficientMemory(offset, self.heap.len()))
+    fn stack_address(
+        &self,
+        size: AddressSize,
+        slot: StackSlot,
+        offset: u64,
+    ) -> Result<Address, MemoryError> {
+        let stack_slots = &self.get_current_function().stack_slots;
+        let stack_slot = &stack_slots[slot];
+
+        // offset must be `0 <= Offset < sizeof(SS)`
+        if offset >= stack_slot.size as u64 {
+            return Err(MemoryError::InvalidOffset {
+                offset,
+                max: stack_slot.size as u64,
+            });
         }
-    }
 
-    fn store_heap(&mut self, offset: usize, v: DataValue) -> Result<(), MemoryError> {
-        if offset + 16 < self.heap.len() {
-            let pointer = self.heap[offset..offset + 16].as_mut_ptr() as *mut _ as *mut u128;
-            Ok(unsafe { v.write_value_to(pointer) })
-        } else {
-            Err(MemoryError::InsufficientMemory(offset, self.heap.len()))
-        }
+        // Calculate offset from the base of the stack, to the current frame
+        let frame_offset: u64 = self
+            .frame_stack
+            .iter()
+            .enumerate()
+            // Remove the last frame, which is the frame that we are currently executing
+            .filter(|&(i, _)| i != self.frame_stack.len() - 1)
+            // Sum all of the stack slots from the previous frames
+            .flat_map(|(_, frame)| frame.function.stack_slots.values())
+            .map(|v| v.size as u64)
+            .sum();
```

Comment:
Would it make sense to track the current sum (basically a "frame pointer") in the interpreter state, rather than compute this dynamically? I could see a worst-case quadratic cost here especially if the fuzzer gets too devious with its test-case generation.